### PR TITLE
Add initial support for gpt-oss

### PR DESCRIPTION
**Pull request by MasterJH5574**
Integrate openai/gpt-oss-20b and gpt-oss-120b into PithTrain: YaRN-scaled RoPE, GQA with interleaved sliding-window / full attention, SwiGLU experts with clamped (alpha, limit) activation and per-expert bias, and the corresponding checkpoint converters and training recipes under examples/. This is the initial gpt-oss integration — further cleanup and optimizations will land in follow-up commits.

Attention sinks are applied as a post-hoc LSE renormalisation on flex_attention's output, not by augmenting the KV sequence with a virtual sink row. Given flex_attention(..., return_lse=True), we multiply attn_output by sigmoid(lse - sinks), which is mathematically identical to softmax over [scores, sink] but avoids a score_mod closure over self.sinks that would otherwise force flex_attention to self-compile and break fullgraph=True on the outer attention region. Same approach as HuggingFace Transformers (flex_attention_forward in integrations/flex_attention.py) and TorchTitan (gpt_oss/model.py).

End-to-end gpt-oss-20b training:
  1. uv venv && uv sync
  2. bash examples/build_tokenized_corpus/launch.sh dclm-gpt-oss
  3. bash examples/pretrain_language_model/launch.sh gpt-oss-20b Checkpoints land in workspace/checkpoints/gpt-oss-20b/ and training resumes from the latest one automatically.

Co-authored-by: Hao Kang <haok1402@gmail.com>

### CODE REVIEW

File: pithtrain/models/gpt_oss.py [hunk 1]
```diff
@@ -0,0 +1,821 @@
...
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
+
+        # YaRN concentration factor (mscale).
+        concentration = 0.1 * math.log(self.scaling_factor) + 1.0
+
+        t = torch.arange(seq_len, device=device, dtype=torch.float32)
+        freqs = torch.outer(t, inv_freq)
+        emb = torch.cat((freqs, freqs), dim=-1)
+        self.register_buffer("cos_cached", (emb.cos() * concentration).to(dtype), persistent=False)
+        self.register_buffer("sin_cached", (emb.sin() * concentration).to(dtype), persistent=False)
```
[the remark this answers is not in the document]
Comment (MasterJH5574) [reply pithtrain/models/gpt_oss.py:127]: incorrect comment.

File: pithtrain/models/gpt_oss.py [hunk 2]
```diff
@@ -0,0 +1,821 @@
...
+            base=rope_theta,
+            scaling_factor=float(rope_scaling.get("factor", 32.0)),
+            original_max_position_embeddings=int(
+                rope_scaling.get("original_max_position_embeddings", 4096)
+            ),
+            beta_fast=float(rope_scaling.get("beta_fast", 32.0)),
+            beta_slow=float(rope_scaling.get("beta_slow", 1.0)),
+        )
+
+        self._block_mask_cache: Dict[int, Tuple[BlockMask, BlockMask]] = {}
```
[the remark this answers is not in the document]
Comment (MasterJH5574) [reply pithtrain/models/gpt_oss.py:717]: Training uses a fixed `seq_len`, so the cache holds at most one entry. Inference in this codebase also goes through DualPipeV at fixed shapes. `BlockMask` stores sparse block-level indices (O((L/block)²)), not a dense L×L tensor, so even a handful of entries would be negligible. Not worth adding LRU plumbing for a speculative case.

File: pithtrain/models/gpt_oss.py [hunk 3]
```diff
@@ -0,0 +1,821 @@
...
+
+class GptOssModel(nn.Module):
+    """GPT-OSS model with DualPipeV pipeline and expert parallelism."""
+
+    def __init__(
+        self,
+        config,
+        num_stages: int,
+        stage_id: int,
+        cp_group: Optional[dist.ProcessGroup] = None,
```
Comment (haok1402) [inline pithtrain/models/gpt_oss.py:632]: `cp_group` is accepted but silently ignored. @MasterJH5574 we should probably fail the CP if used.

File: pithtrain/models/gpt_oss.py [hunk 4]
```diff
@@ -0,0 +1,821 @@
...
+    """
+    Grouped Query Attention with attention sinks and optional sliding window.
+
+    Attention sinks are a learned per-head scalar added to the softmax
+    denominator.  We apply them as a post-hoc LSE renormalisation on the
+    flex_attention output rather than via a ``score_mod`` closure, so that
+    the surrounding ``_forward_attn_compute`` stays ``@torch.compile(fullgraph=True)``
+    traceable.  Large sink values let a head "dump" attention mass to the
+    sink, producing near-zero attention to real tokens.  See the forward
+    method for the renorm math and ``docs/gpt_oss_lse_renorm_brief.md`` for
```
Comment (haok1402) [inline pithtrain/models/gpt_oss.py]: `docs/gpt_oss_lse_renorm_brief.md` was never included in the commit. @MasterJH5574 We should probably drop the reference.

File: pithtrain/models/gpt_oss.py [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
+
+        # Attention sinks via post-hoc LSE renormalisation (HF / TorchTitan
+        # recipe).  The alternative — a score_mod closure over self.sinks with
+        # a virtual zero-valued KV row — forces flex_attention to self-compile
+        # and prevents the outer @torch.compile(fullgraph=True) from tracing
+        # through, which leaves the whole attention block in eager mode.  The
+        # renormalisation here is mathematically identical:
+        #   softmax_with_sink(s)_i = exp(s_i) / (sum_j exp(s_j) + exp(sink))
+        #                          = softmax(s)_i * sigmoid(lse − sink)
+        # See docs/gpt_oss_lse_renorm_brief.md for the full derivation.
```
Comment (haok1402) [inline pithtrain/models/gpt_oss.py]: Same here ; `docs/gpt_oss_lse_renorm_brief.md` was never committed.

File: pithtrain/models/gpt_oss.py [hunk 6]
```diff
@@ -0,0 +1,821 @@
...
+from pithtrain.operators.token_scatter import padded_index_gather, scatter_for_grouped_gemm
+
+torch._dynamo.allow_in_graph(MoELoadBalanceLossInjector)
+
+# ---------------------------------------------------------------------------
+# YaRN Rotary Position Embedding
+# ---------------------------------------------------------------------------
+
+SWIGLU_ALPHA = 1.702
+SWIGLU_LIMIT = 7.0
```
Comment (haok1402) [inline pithtrain/models/gpt_oss.py]: `swiglu_limit` was a value from the `config.json` file with gpt-oss. @MasterJH5574 Perhaps the best if we could read from the config and not declared as a module-level constant.

File: pithtrain/tasks/convert_checkpoint.py [hunk 7]
```diff
@@ -87,10 +216,50 @@ def hf2dcp(cfg: ConvertCheckpointCfg, stdout: Logger) -> None:
...
+
+
+def _stack_experts(state_dict: Dict[str, torch.Tensor], stdout: Logger) -> Dict[str, torch.Tensor]:
+    """Stack per-expert canonical keys along dim 0 and transpose 3-D weights
+    to HF's live ``[E, in, out]`` layout.  Our DCP stores ``[E, out, in]``
+    (TorchTitan / TE convention); the transpose lives at this HF boundary
+    so the model and ``hf2dcp`` never need to see it.  1-D biases pass
+    through unchanged.
     """
+    import re
```
Comment (haok1402) [inline pithtrain/tasks/convert_checkpoint.py]: nit : this can be imported at the module-level.

File: pithtrain/models/gpt_oss.py [hunk 8]
```diff
@@ -0,0 +1,821 @@
...
+
+        gate_up = self._grouped_mm(x, self.gate_up_proj, grouped_mm_offs)
+        gate_up = gate_up + self.gate_up_proj_bias[group_ids]
+        gate = gate_up[:, ::2]
+        up = gate_up[:, 1::2]
+
+        gate = gate.clamp(max=SWIGLU_LIMIT)
+        up = up.clamp(min=-SWIGLU_LIMIT, max=SWIGLU_LIMIT)
+        glu = gate * torch.sigmoid(SWIGLU_ALPHA * gate)
+        activated = (up + 1) * glu
```
Comment (haok1402) [inline pithtrain/models/gpt_oss.py]: Like `swiglu`, activation here might benefit from recompute in a fused kernel @MasterJH5574

File: pithtrain/models/gpt_oss.py (hunk 3 above)
Comment (MasterJH5574) [reply pithtrain/models/gpt_oss.py:632]: Agreed.  Raising error now if cp_group is defined and group size is more than 1.

File: pithtrain/models/gpt_oss.py (hunk 4 above)
Comment (MasterJH5574) [reply pithtrain/models/gpt_oss.py]: Thanks for catching this.  Removed now.

File: pithtrain/models/gpt_oss.py (hunk 6 above)
Comment (MasterJH5574) [reply pithtrain/models/gpt_oss.py]: Thanks for catching! Now referring to the config file for the limit value.

File: pithtrain/tasks/convert_checkpoint.py (hunk 7 above)
Comment (MasterJH5574) [reply pithtrain/tasks/convert_checkpoint.py]: Resolved.

File: pithtrain/models/gpt_oss.py (hunk 3 above)
Comment (haok1402) [reply pithtrain/models/gpt_oss.py:632]: Yes ; we're yet to implement CP support for gpt-oss.

File: pithtrain/models/gpt_oss.py (hunk 8 above)
Comment (MasterJH5574) [reply pithtrain/models/gpt_oss.py]: Yes, this is a planned followup. Will do that in a PR later.

**Comment by haok1402** [comment]
Right now the model checkpoint conversion only runs with a single process. For gpt-oss-120b, we see about 500 GiB CPU RAM usage, which can OOM for even larger ones. Might need a better design with this.

```bash
haok@orchard-flame-31:~/Pith-Train$ ls -lha workspace/checkpoints/gpt-oss-120b/torch-dcp/step-00000000/
total 218G
drwxr-xr-x 2 haok users 4.0K Apr 18 21:36 .
drwxr-xr-x 3 haok users 4.0K Apr 18 21:32 ..
-rw-r--r-- 1 haok users 5.2M Apr 18 21:36 .metadata
-rw-r--r-- 1 haok users 218G Apr 18 21:36 __0_0.distcp
```

**Comment by haok1402** [comment]
The cross entropy loss starts off 5.74 from the released checkpoint is too high for a properly pretrained gpt-oss-20b. Expected the loss to starts off in the range of 2 - 3ish. 

```
2026-04-18 22:03:09 | INFO | step 00000001/00000025 | step-time 83.551 sec | cross-entropy-loss 5.7470 | load-balance-loss 1.612387 | learning-rate 1.000000e-06 | gradient-norm 563.5688 | tokens-per-second 25,100 | peak-gpu-memory 57.42 GB
2026-04-18 22:03:44 | INFO | step 00000002/00000025 | step-time 34.556 sec | cross-entropy-loss 5.9446 | load-balance-loss 1.621795 | learning-rate 1.000000e-06 | gradient-norm 537.4495 | tokens-per-second 60,689 | peak-gpu-memory 57.36 GB
2026-04-18 22:04:20 | INFO | step 00000003/00000025 | step-time 35.779 sec | cross-entropy-loss 5.7348 | load-balance-loss 1.599342 | learning-rate 1.000000e-06 | gradient-norm 680.6374 | tokens-per-second 58,614 | peak-gpu-memory 57.42 GB
2026-04-18 22:04:55 | INFO | step 00000004/00000025 | step-time 34.846 sec | cross-entropy-loss 5.8458 | load-balance-loss 1.612310 | learning-rate 1.000000e-06 | gradient-norm 306.0502 | tokens-per-second 60,183 | peak-gpu-memory 57.42 GB
2026-04-18 22:05:30 | INFO | step 00000005/00000025 | step-time 34.332 sec | cross-entropy-loss 5.6770 | load-balance-loss 1.618142 | learning-rate 1.000000e-06 | gradient-norm 447.3632 | tokens-per-second 61,084 | peak-gpu-memory 57.47 GB
2026-04-18 22:06:04 | INFO | step 00000006/00000025 | step-time 34.385 sec | cross-entropy-loss 5.4798 | load-balance-loss 1.594669 | learning-rate 1.000000e-06 | gradient-norm 423.0066 | tokens-per-second 60,991 | peak-gpu-memory 57.45 GB
```

File: pithtrain/models/gpt_oss.py [hunk 9]
```diff
@@ -0,0 +1,831 @@
...
+        2 * math.log(base)
+    )
+
+
+def _yarn_find_correction_range(
+    low_rot: float, high_rot: float, dim: int, base: float, max_position_embeddings: int
+) -> Tuple[int, int]:
+    low = math.floor(_yarn_find_correction_dim(low_rot, dim, base, max_position_embeddings))
+    high = math.ceil(_yarn_find_correction_dim(high_rot, dim, base, max_position_embeddings))
+    return max(low, 0), min(high, dim - 1)
```
Comment (haok1402) [inline pithtrain/models/gpt_oss.py]: @MasterJH5574 We shouldn't applied the floor and ceil by default. The `rope_scaling` for the model config contains a `truncate` key, which indicates whether or not the floor + ceil should be applied.

- https://huggingface.co/openai/gpt-oss-20b/blob/main/config.json 
- https://huggingface.co/openai/gpt-oss-120b/blob/main/config.json

File: examples/pretrain_language_model/gpt-oss-120b/config.json [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  "num_local_experts": 128,
+  "pad_token_id": 199999,
+  "rms_norm_eps": 1e-05,
+  "rope_scaling": {
+    "beta_fast": 32.0,
+    "beta_slow": 1.0,
+    "factor": 32.0,
+    "original_max_position_embeddings": 4096,
+    "rope_type": "yarn"
+  },
```
Comment (haok1402) [inline examples/pretrain_language_model/gpt-oss-120b/config.json:48]: Missing the `truncate` field in `rope_scaling` model config.

File: examples/pretrain_language_model/gpt-oss-20b/config.json [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  "num_local_experts": 32,
+  "pad_token_id": 199999,
+  "rms_norm_eps": 1e-05,
+  "rope_scaling": {
+    "beta_fast": 32.0,
+    "beta_slow": 1.0,
+    "factor": 32.0,
+    "original_max_position_embeddings": 4096,
+    "rope_type": "yarn"
+  },
```
Comment (haok1402) [inline examples/pretrain_language_model/gpt-oss-20b/config.json:42]: Missing the `truncate` field in `rope_scaling` model config.

File: pithtrain/models/gpt_oss.py (hunk 9 above)
Comment (MasterJH5574) [reply pithtrain/models/gpt_oss.py]: Addressed.
